examples/sync/src/protocol.rs

```diff
@@ -0,0 +1,389 @@
+//! Network protocol definitions for ADB sync.
+//!
+//! This module defines the network protocol used for ADB synchronization between
+//! clients and servers. It includes message types, error handling, and validation
+//! logic for safe network communication.
+//!
+//! The protocol supports:
+//! - Getting server metadata (database size, target hash, operation bounds)
+//! - Fetching operations with cryptographic proofs
+//! - Comprehensive error handling
+
+use bytes::{Buf, BufMut};
+use commonware_codec::{EncodeSize, Error as CodecError, Read, ReadExt, ReadRangeExt as _, Write};
+use commonware_cryptography::sha256::Digest;
+use std::num::NonZeroU64;
+use thiserror::Error;
+
+/// Maximum message size in bytes (10MB).
+pub const MAX_MESSAGE_SIZE: usize = 10 * 1024 * 1024;
+
+/// Network protocol messages for ADB sync.
+#[derive(Debug, Clone)]
+pub enum Message {
+    /// Request operations from the server.
+    GetOperationsRequest(GetOperationsRequest),
+    /// Response with operations and proof.
+    GetOperationsResponse(GetOperationsResponse),
+    /// Request server metadata (target hash, bounds, etc.).
+    GetServerMetadataRequest,
+    /// Response with server metadata.
+    GetServerMetadataResponse(GetServerMetadataResponse),
+    /// Error response.
+    Error(ErrorResponse),
+}
+
+/// Request for operations from the server.
+#[derive(Debug, Clone)]
+pub struct GetOperationsRequest {
+    /// Size of the database at the root we are syncing to.
+    pub size: u64,
+    /// Starting location for the operations.
+    pub start_loc: u64,
+    /// Maximum number of operations to return.
+    pub max_ops: NonZeroU64,
+}
+
+/// Response with operations and proof.
+#[derive(Debug, Clone)]
+pub struct GetOperationsResponse {
+    /// Serialized proof that the operations were in the database.
```

Comment:
We should parse both of these during `read_cfg`? Not sure why we'd expose the raw bytes.

Reply:
We should be able to provide any CFG needed for parsing via `type = Cfg`.